tbmalt/physics/dftb/__init__.py

```diff
@@ -1067,238 +1070,445 @@ def forward(self, cache: Optional[Dict[str, Any]] = None
-
-            The charges ``q_in`` and ``q_out`` may be either shell or atom
-            resolved, but must match up with that as defined by the orbs
-            attribute `shell_resolved`.
+            q_out: the resulting output charges.
+            hamiltonian: the second order hamiltonian matrix. When ``q_in`` is
+                equal to ``q_out`` this will be the "true" self-consistent
+                charge hamiltonian matrix.
+            eig_values: corresponding eigenvalues.
+            eig_vectors: associated eigenvectors.
+            rho: density matrix.
         """
-
         # Construct the shift matrix
         shifts = torch.einsum(
-            '...i,...ij->...j', q_in - self.q_zero_res, self.gamma)
-        shifts = prepeat_interleave(shifts, self.orbs.orbs_per_res)
+            '...i,...ij->...j', q_in - q_zero, gamma)
+
+        shifts = prepeat_interleave(shifts, orbs.orbs_per_res)
         shifts = (shifts[..., None] + shifts[..., None, :])
 
         # Compute the second order Hamiltonian matrix
-        self._hamiltonian = self.core_hamiltonian + .5 * self.overlap * shifts
+        hamiltonian = core_hamiltonian + .5 * overlap * shifts
 
         # Obtain the eigen-values/vectors via an eigen decomposition
-        self.eig_values, self.eig_vectors = eighb(
-            self.hamiltonian, self.overlap, **self._solver_settings)
+        eigen_solver_settings = kwargs.get('eigen_solver_settings', {})
+        eig_values, eig_vectors = eighb(hamiltonian, overlap, **eigen_solver_settings)
+
+        # Calculate Occupancies
+        # ---------------------
+        # Note that this scale factor assumes spin-restricted and will need to
+        # be refactored when implementing spin-unrestricted calculations.
+        scale_factor = 2.0
+
+        e_mask = orbs if orbs.atomic_numbers.dim() == 2 else None
+
+        # Compute the new occupancy values, performing smearing as necessary.
+        if filling_scheme is not None:
+            fermi_energy = fermi_search(eig_values, n_electrons, filling_temp,
+                                        filling_scheme, e_mask=e_mask)
+
+            occupancy = filling_scheme(eig_values, fermi_energy, filling_temp,
+                                       e_mask=e_mask) * scale_factor
+
+        else:
+            occupancy = aufbau_filling(eig_values, n_electrons,
+                                       e_mask=e_mask) * scale_factor
 
         # Scaled occupancy values
         s_occs = torch.einsum(
-            '...i,...ji->...ji', torch.sqrt(self.occupancy), self.eig_vectors)
+            '...i,...ji->...ji', torch.sqrt(occupancy), eig_vectors)
 
         # Density matrix
-        self.rho = s_occs @ s_occs.transpose(-1, -2).conj()
+        rho = s_occs @ s_occs.transpose(-1, -2).conj()
+
+        # Compute the new charges
+        q_out = _mulliken(rho, overlap, orbs)
+
+        return q_out, hamiltonian, eig_values, eig_vectors, rho
+
+    # @staticmethod
+    # def scc_cycle_x(
```

Comment:
Hey Adam, seems this function was used temporarily for debugging?

Reply:
Yes, you are right. I am not sure how I managed to miss that.